src/MongoDB.Driver.Core/Core/Servers/Server.cs

```diff
@@ -233,6 +228,13 @@ public void RequestHeartbeat()
             _monitor.RequestHeartbeat();
         }
 
+        private bool ContainsTimeOutException(MongoConnectionException connectionException)
+        {
+            return connectionException.InnerException is IOException ioException &&
+                   ioException.InnerException is SocketException socketException &&
+                   socketException.SocketErrorCode == SocketError.TimedOut;
+        }
```

Comment:
Optional suggestion: replace this method with a less rigid implementation:

    private bool ContainsSocketTimeOutException(Exception exception)
    {
        while (exception != null)
        {
            if (exception is SocketException socketException && socketException.SocketErrorCode == SocketError.TimedOut)
            {
                return true;
            }

            exception = exception.InnerException;
        }

        return false;
    }

This alternative will find the timeout `SocketException` no matter how deep it is or what exceptions the path to it goes through.

Reply:
Done